cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/BasicLineageWriter.java

```diff
@@ -102,4 +104,26 @@ public void addAccess(ProgramRunId run, StreamId streamId, AccessType accessType
   protected DatasetId getLineageDatasetId() {
     return LineageDataset.LINEAGE_DATASET_ID;
   }
+
+  /**
+   * Returns the {@link DatasetId} of the field lineage dataset. This method should only be overridden in unit-test.
+   */
+  @VisibleForTesting
+  protected DatasetId getFieldLineageDatasetId() {
+    return FieldLineageDataset.FIELDLINEAGE_DATASET_ID;
+  }
+
+  @Override
+  public void write(ProgramRunId programRunId, FieldLineageInfo info) {
+    Transactionals.execute(transactional, context -> {
+      FieldLineageDataset fieldLineageDataset
+        = FieldLineageDataset.getFieldLineageDataset(context, datasetFramework, getFieldLineageDatasetId());
+
+      if (!fieldLineageDataset.hasFieldLineageInfo(info.getChecksum())) {
```

Comment:
I feel that this logic (adding the checksum row if necessary, then adding the reference rows) should be encapsulated in the dataset's API. In other words, this should be a single method call here.